compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -866,6 +866,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 // be peeled to `str` while ty here is still `&str`, if we don't
                 // err early here, a rather confusing unification error will be
                 // emitted instead).
+                let ty = self.structurally_resolve_type(expr.span, ty);
```

Comment:
do we need to do the same in `fn calc_adjust_mode` given that this also structurally matches on `check_pat_expr_unadjusted`?

Reply:
possibly